src/tools/tests/schema2graph-test.ts

```diff
@@ -593,4 +593,129 @@ describe('schema2graph', () => {
+      {name: 'T_H1', parents: '', children: ''},
+      {name: 'T_H2', parents: '', children: ''},
+    ]);
+    assert.deepStrictEqual(res.aliases, {
+      'T_H1': ['T_H1'],
+      'T_H2': ['T_H2'],
+    });
+  });
+
+  it('variables with no constraints are not empty', async () => {
+    const manifest = await Manifest.parse(`
+      particle T
+        h1: reads ~a                               // 1 
+        h2: reads ~a                               // 1
+    `);
+    const res = convert(new SchemaGraph(manifest.particles[0]));
+    assert.deepStrictEqual(res.nodes, [
+      {name: 'T_H1', parents: '', children: ''},
+    ]);
+    assert.deepStrictEqual(res.aliases, {
+      'T_H1': ['T_H1', 'T_H2'],
+    });
+  });
+
+  it('distinct variables with the same constraint should be distinct', async () => {
+    const manifest = await Manifest.parse(`
+      particle T
+        h1: reads ~a with {foo: Text}               // 1 
+        h2: reads ~b with {foo: Text}               // 2 
+    `);
+    const res = convert(new SchemaGraph(manifest.particles[0]));
+    assert.deepStrictEqual(res.nodes, [
+      {name: 'T_H1', parents: '', children: ''},
+      {name: 'T_H2', parents: '', children: ''},
+    ]);
+    assert.deepStrictEqual(res.aliases, {
+      'T_H1': ['T_H1'],
+      'T_H2': ['T_H2'],
+    });
+  });
+
+  it('variables are aliased correctly within a scope', async () => {
+    // Type variables with the same name in a particle scope should be treated
+    // as the same type, even with constraints applied.
+    // Variables with different names but the same constraints should be distinct.
+    const manifest = await Manifest.parse(`
+      particle T
+        h1: reads ~a with {a: Text}                 // 1 
+        h2: reads ~b                                // 2
+        h3: reads ~c                                // 3
+        h4: reads ~d with {a: Text}                 // 4
+        h5: writes ~a                               // 1
+        h6: writes ~b with {b: Number}              // 2
+        h7: writes ~c                               // 3 
+        h8: writes ~d                               // 4
+    `);
+    const res = convert(new SchemaGraph(manifest.particles[0]));
+    assert.deepStrictEqual(res.nodes, [
+      {name: 'T_H1', parents: '', children: ''},
+      {name: 'T_H2', parents: '', children: ''},
+      {name: 'T_H3', parents: '', children: ''},
+      {name: 'T_H4', parents: '', children: ''},
+    ]);
+    assert.deepStrictEqual(res.aliases, {
+      'T_H1': ['T_H1', 'T_H5'],
+      'T_H2': ['T_H2', 'T_H6'],
+      'T_H3': ['T_H3', 'T_H7'],
+      'T_H4': ['T_H4', 'T_H8'],
+    });
+  });
+
+  it('type variables aggregate all constraints', async () => {
+    const manifest = await Manifest.parse(`
+      particle T
+        h1: reads ~a with {a: Text}                     // [x] 1 -> 2 -> 3
+        h2: reads ~a with {a: Text, b: Text}            // [√] 1
+        h3: reads ~a with {a: Text, b: Text, c: Text}
```

Comment:
what if you remove `a` from here, then we can be check that the resulting schema will indeed aggregate constraints instead of just using the last one.